pystan/diagnostics.py

```diff
@@ -0,0 +1,389 @@
+    try:
+        n_effs_index = fit_summary['summary_colnames'].index('n_eff')
+    except:
+        raise ValueError('Summary of fit object appears to lack information on effective sample size')
+        
+    n_effs = [x[n_effs_index] for x in fit_summary['summary']]
+    names = fit_summary['summary_rownames']
+    n_iter = len(fit.extract()['lp__'])
+
+    no_warning = True
+    for n_eff, name in zip(n_effs, names):
+        ratio = n_eff / n_iter
+        if (ratio < 0.001):
+            if verbosity > 0:
+                print('n_eff / iter for parameter {} is {}!'.format(name, ratio))
+                
+            no_warning = False
+            
+    if no_warning:
+        if verbosity > 1:
+            print('n_eff / iter looks reasonable for all parameters')
+
+        return True
+    else:
+        if verbosity > 0:
+            print('n_eff / iter below 0.001 indicates that the effective sample size has likely been overestimated')
+
+def check_rhat(fit, verbose = True):
+    """Checks the potential scale reduction factors, i.e., Rhat values
+
+    Parameters
+    ----------
+    fit : StanFit4Model object
+    verbose : bool or int, optional
+        If ``verbose`` is ``False`` or a nonpositive integer, no
+        diagnostic messages are printed, and only the return value of
+        the function conveys diagnostic information. If it is ``True``
+        (the default) or an integer greater than zero, then a
+        diagnostic message is printed only if there are Rhat values
+        too far from 1. If ``verbose`` is an integer greater than 1,
+        then extra diagnostic messages are printed.
+
+
+    Returns
+    -------
+    bool
+        ``True`` if there are no problems with with Rhat and ``False``
+        otherwise.
+
+    Raises
+    ------
+    ValueError
+        If the output of ``fit.summary()`` has no information about
+        Rhat.
+
+    """
+    
+    from math import isnan
+    from math import isinf
+
+    verbosity = int(verbose)
+
+    fit_summary = fit.summary(probs=[0.5])
+
+    try:
+        Rhat_index = fit_summary['summary_colnames'].index('Rhat')
+    except:
+        raise ValueError('Summary of fit object appears to lack information on Rhat')
+    
+    rhats = [x[Rhat_index] for x in fit_summary['summary']]
+    names = fit_summary['summary_rownames']
+
+    no_warning = True
+    for rhat, name in zip(rhats, names):
+        if (isnan(rhat) or isinf(rhat) or rhat > 1.1 or rhat < 0.9):
+
+            if verbosity > 0:
+                print('Rhat for parameter {} is {}!'.format(name, rhat))
+                
+            no_warning = False
+            
+    if no_warning:
+        if verbosity > 1:
+            print('Rhat looks reasonable for all parameters')
+
+        return True
+    else:
+        if verbosity > 0:
+            print('Rhat above 1.1 or below 0.9 indicates that the chains very likely have not mixed')
+
+        return False
+
+def check_MCMC_diagnostics(fit, verbose = True):
```

Comment:
maybe name this just as `check_diagnostics` and then check that the sampler was nuts (or hmc).
Because later we could add some diagnostics for `vb` (if they come mainstraim).

Reply:
I think that *for now* it may be best to leave the name as `check_MCMC_diagnostics`, since

1. the name implies (correctly) that it is specifically for fit objects generated from MCMC sampling, and

2. it allows the function to have MCMC-specific options, such as an argument indicating if per-chain diagnostics are desired.

A future diagnostics function for `vb` could always be called `check_vb_diagnostics()`.

Reply:
Yes, that sounds a good option.